library/std/src/sys_common/time.rs

```diff
@@ -0,0 +1,100 @@
+#[cfg_attr(not(windows), allow(dead_code))]
+const MONOTONIZE_U64: bool = cfg!(target_has_atomic = "64");
+
+#[cfg_attr(not(windows), allow(dead_code))]
+pub fn monotonize_u64(tick: u64) -> Option<u64> {
+    if MONOTONIZE_U64 {
+        use crate::sync::atomic::AtomicU64;
+        use crate::sync::atomic::Ordering::Relaxed;
+
+        static LAST_NOW: AtomicU64 = AtomicU64::new(0);
+
+        return Some(LAST_NOW.fetch_max(tick, Relaxed));
+    }
+
+    None
+}
+
+#[cfg_attr(not(any(unix, target_os = "ios", target_os = "macos")), allow(dead_code))]
+pub fn can_monotonize_u64x2() -> bool {
+    if cfg!(target_arch = "x86_64") {
+        #[cfg(not(test))]
+        use crate::detect::arch::is_x86_feature_detected;
+        #[cfg(test)]
+        use crate::realstd::is_x86_feature_detected;
+
+        return is_x86_feature_detected!("cmpxchg16b");
+    }
+
+    false
+}
+
+#[cfg_attr(not(any(unix, target_os = "ios", target_os = "macos")), allow(dead_code))]
+pub fn monotonize_u64x2(sec: u64, nsec: u64) -> Option<(u64, u64)> {
+    if can_monotonize_u64x2() {
+        let os_now = (sec as u128) << 64 | nsec as u128;
+        let monotonized = unsafe { monotonize_cmpxchg16b(os_now) };
+        return Some(((monotonized >> 64) as u64, monotonized as u64));
+    }
+
+    None
+}
+
+
+#[cfg_attr(not(any(unix, target_os = "ios", target_os = "macos")), allow(dead_code))]
+#[cfg(target_arch = "x86_64")]
+#[target_feature(enable = "cmpxchg16b")]
+// FIXME(#56798) inlining this on baseline x86_64 would fail to link __sync_val_compare_and_swap_16
+#[cfg_attr(not(target_feature = "cmpxchg16b"), inline(never))]
+unsafe fn monotonize_cmpxchg16b(os_now: u128) -> u128 {
```

Comment:
Could you use `AtomicU128::fetch_add` instead? It is available on x86_64 macOS: https://github.com/rust-lang/rust/blob/9b6339e4b9747d473270baa42e77e1d2fff39bf4/compiler/rustc_target/src/spec/x86_64_apple_darwin.rs#L6 The `cmpxchgb16` target feature is also enabled by default on macOS as the target cpu is `core2`.

Reply:
macOS already has `Instant::actually_monotonic() == true`, so it wouldn't need it anyway.
But using AtomicU128 is still be a good idea for aarch64 ports of other unixes.